### ISSUE TITLE
project maintenance based on rom

### ISSUE THREAD

**Issue by 61t** [josiahcarlson/rom#117]
Thank you @josiahcarlson . I got the project first, then after checked your github repository. I was aware that you are the author of Redis in Action, that's definitely best book on Redis at least for application in different fields with details and thoughts. Glad to see your next edition for Redis in Action. 

I am planning to use rom in my project with Redis. After went through your docs, it is still not clear what should we do about feature/model change with rom. 

My Concerns (03/06/2018 updated): 
1. adding/removing new index 
2. adding/removing cols 
3. session mgmt for perf 
4. lower level integration with rom/redis

My Thoughts: 
1. Adding is covered with rom.util.refresh_indices, how about removing/altering. 
2. Any concern? rom.util.refresh_cols with default value? 
3. This is covered with rom.util.Session, I will unit test it today. 
4. I believe rom is sqlalchemy orm, and redis-py is sqlalchemy core for now. As before go through rom source code there is no clue about key pattern and data dependency for rom. I think I should go through source code first, and any reference? 

At last thanks again. If I would like recommend a book about Redis with python for application in real world, Redis in Action is the best option or might be the only option for now 👍

**Comment by josiahcarlson** [comment]
I am currently on vacation with limited internet connectivity. I will
answer your questions when I get back in about a week.

**Comment by 61t** [comment]
@josiahcarlson No hurry at all. Have fun and enjoy the vacation. :)

**Comment by josiahcarlson** [comment]
> Thank you @josiahcarlson . I got the project first, then after checked your github repository. I was aware that you are the author of Redis in Action, that's definitely best book on Redis at least for application in different fields with details and thoughts. Glad to see your next edition for Redis in Action.

If you mean the Simplified Chinese translation, then yes, I am very happy to see it too! I haven't seen it in person yet, but I hope to get a copy in the mail soon. :D

> I am planning to use rom in my project with Redis. After went through your docs, it is still not clear what should we do about feature/model change with rom.
> 
> My Concerns (03/06/2018 updated):
> 
> adding/removing new index



> adding/removing cols

Adding: just add the new columns. Note that strange behavior may occur if you use "`required=true`" in combination with new columns. I'd just have sane "null" values on new columns until you've backfilled in some default data, then change the default on your model. I've done this literally dozens of times in my own projects, professional, personal, and otherwise. Works pretty well.

Removing: Set "`default=None, required=False`" on a column you want to remove, slowly remove old data. When all old data is out, remove the column from the model. Also done this dozens of times. Works pretty well.

> session mgmt for perf

Fewer round-trips = better, generally. So the current session manager keeps changes until you do "`session.commit()`" You can disable this, and all caching, with documented defaults. Check the docs in English here: http://pythonhosted.org/rom/rom.html#using-a-non-caching-session-object

> lower level integration with rom/redis

How low-level do you want to go?

> My Thoughts:
> 
> Adding is covered with rom.util.refresh_indices, how about removing/altering.

It is. Also, as an alternate interface: http://pythonhosted.org/rom/rom.html#rom.util.clean_old_index

> Any concern? rom.util.refresh_cols with default value?

Yeah, that might be funky. I'd wait to add a "`default=val`` if `val != None`, until you'd backfilled the data.

> This is covered with rom.util.Session, I will unit test it today.

Mostly, yes.

> I believe rom is sqlalchemy orm, and redis-py is sqlalchemy core for now. As before go through rom source code there is no clue about key pattern and data dependency for rom. I think I should go through source code first, and any reference?

Rom is basically the applied implementation of chapter 7 of my book, minus the ad stuff. Indeed, it is meant to cover many/most use-cases of SQLAlchemy ORM, inside Redis, minus joins. Though foreign key constraints *are* held, and the constraints are also used as indexes, like a traditional database. Redis is a database construction kit, and I've just built a database minus aggregations (missing a query language, for now) on top of it. Mostly data is stored a json dump of data. You can have sub-object columns (json object columns storing lists of things, indexed in certain ways), but not direct access to underlying structures themselves.

Unlike other Redis database interfaces, we generally don't offer direct access to intersections, unions, differences, etc., but build higher-level concepts on top of them. Data usage is generally moderate to high, and some queries on large sets can be slow; but that's a Redis problem, not a rom problem.

At some point it may make sense to re-implement with modules in mind, but if I can't pay someone to host a Redis module, I'm not interested in implementing to that spec, and I can't expect anyone else to either. I like the idea, I just think modules should be in Lua (or at least not C/C++ due to the build process nightmares I've personally experienced).

If you want to see how data maps between Python and Redis, check: https://github.com/josiahcarlson/rom/blob/master/rom/columns.py

If you want to see how data is indexed/queried, check: https://github.com/josiahcarlson/rom/blob/master/rom/index.py and https://github.com/josiahcarlson/rom/blob/master/rom/query.py , respectively.

> At last thanks again. If I would like recommend a book about Redis with python for application in real world, Redis in Action is the best option or might be the only option for now +1

I am happy to have made something that helps solve problems. If you are planning on using this in your corporation, you may want to consider hiring me at least temporarily as a consultant. Not only will you get private (instead of public support), I can give you possibly the best advice on how to scale your Redis infrastructure, as well as the data operations you will be performing (whether with rom or some other Redis package). My email address is on my public Github profile.

Let me know if I can answer any more questions.

**Comment by 61t** [comment]
@josiahcarlson welcome back. Had a wonderful vacation didn't you :) 

As for your book, I mean it is the best one in English world at least for me. Most of redis books explain what we can easily find from official docs.  Why not reading directly from official docs first, is it? 

Redis official docs are math reference books, and Redis in Action is a book about Applied Math to deal with issues in real world. That is my point. 

"required=true" concerns and removing columns sound reasonable. I personally prefer numpy's missing data strategy as NaN or 1, and sparse ndarray concept in math, so "required=true" is no problem at all for me, and back-filed operation is more about hist data and normalization. 

Session mgmt for insert/update performance works very well. But delete perf is bad. Once I got a timeout alert for deleting 300K entities (As my deleting/paging out job didn't run at that time), but didn't finish it within 1 minute. After looking inside rom, there is only one way to delete entities one by one (300K round-trip inside, even iter_result didn't help here as I tested). If we can merge deleting action inside session mgmt, that will be awesome. I would like to contribute this feature if I can help. 

Low level integration with redis-py was answered below with the link about data map between rom and redis. I will check it this week. 

Query features as I tested are very mature for almost all of scenarios. If one didn't agree with it and think of more fancy query feature based on experience from other lib or db, one should redesign their model and dataflow. 

The meaning of Low level integration for me is more about timing critical data operation rather than db operation (S-I-U-D), and I totally agree with "Data usage is generally moderate to high" and I would like to add a comment "If it has to be low in your dataflow, you should consider rom as data-input of math lib rather than of math operation by hand yourself". 

Currently I am more focusing on my personal project, and it might turn to a biz in the future. If I need consultant of redis, you are no-doubt the 1st one in my mind. You rocks. :)

**Comment by josiahcarlson** [comment]
> @josiahcarlson welcome back. Had a wonderful vacation didn't you :)

Being home is more relaxing, but I have young kids. :P

> As for your book, I mean it is the best one in English world at least for me. Most of redis books explain what we can easily find from official docs. Why not reading directly from official docs first, is it?
> 
> Redis official docs are math reference books, and Redis in Action is a book about Applied Math to deal with issues in real world. That is my point.

Ahh, yes. Definitely the entire point of the book. Without examples of things being used to solve problems, it's hard to see the point. :)

> Session mgmt for insert/update performance works very well. But delete perf is bad. Once I got a timeout alert for deleting 300K entities (As my deleting/paging out job didn't run at that time), but didn't finish it within 1 minute. After looking inside rom, there is only one way to delete entities one by one (300K round-trip inside, even iter_result didn't help here as I tested). If we can merge deleting action inside session mgmt, that will be awesome. I would like to contribute this feature if I can help.

I'm not sure how good insert/update really is. Internally, all objects are manipulated one at a time. And even creating a new object gets a round-trip to Redis (due to the automatic creation of the PrimaryKey() value). With a little work, deleting could be made fast by half breaking the index and cleaning it up, but I suppose just doing a bulk delete on the session with `session.delete()` like the existing `session.save()` (though `session.save()` does at least 1 round-trip per entity saved, fyi), and optimizing it with a single round-trip would probably go pretty far.

Did you have a different API in mind than `session.delete()`?

> Query features as I tested are very mature for almost all of scenarios. If one didn't agree with it and think of more fancy query feature based on experience from other lib or db, one should redesign their model and dataflow.

It's funny, but my experience with Redis and relational databases brought me to rom as about 90-95% of what I use for everything except for analysis. I'm slowly finishing rom to get that last 5-10% of my needs, and hopefully more of everyone else's too. :)

> The meaning of Low level integration for me is more about timing critical data operation rather than db operation (S-I-U-D), and I totally agree with "Data usage is generally moderate to high" and I would like to add a comment "If it has to be low in your dataflow, you should consider rom as data-input of math lib rather than of math operation by hand yourself".

Definitely. Right now that is a limitation with Redis itself, which is slowly getting better (there were some recent data optimizations that Redis has done recently). There are 2-3 different ways I can address it in the future, and I'm working on one of them right now. Hope to release soon. :)

**Comment by 61t** [comment]
Awesome reply. I do have some thoughts in my mind about deleting. I will get back here tomorrow with some test results I have done. Time to get off work. Have a good night :)

**Comment by 61t** [comment]
I got my previous test code: 

`
r = redis.StrictRedis(host=host, port=port)

del_num = r.delete(*r.keys())

print('{} Keys deleted'.format(del_num))
`

Above code with redis-py will easily handle xxxK at least data in several seconds as I remembered .

**Comment by josiahcarlson** [comment]
Well, that clears everything and basically resets your whole database. That's not really viable for real use-cases where "destroy the world" is usually a bad thing. I'm talking about being able to do something like...

```
de = []
chunk = 100
for i, result in MyModel.query.filter(...).iter_result(pagesize=chunk):
    if i and not i % chunk:
        session.delete(de) # one round-trip to delete "chunk" items
        del de[:]
    de.append(result)
if de:
    session.delete(de)
```

So to make deletion of data reasonable (not just database resets), `session.delete(list_or_iterable)` plus adding a `.delete()` method to a query object, which does the above for configurable `block` sizes. Might get a bit nasty with foreign key relationships, but I'm probably just going to not support that for now.

I've got an implementation of this already, because it's something that I've needed too.

**Comment by 61t** [comment]
@josiahcarlson reset whole test db is on purpose as part of pytest. It is not viable at all for production for sure as you said. Your code snippet is exactly what I mean and the way I am thinking. 

Session.delete and Query.delete are beautiful solution as you mentioned. I didn't see it in rom 0.40.2, features in next commit ?

**Comment by josiahcarlson** [comment]
Yes, these delete features will be in the next release, which should be expected in the next week or so.

**Comment by 61t** [comment]
Awesome Thank you very much.

**Comment by 61t** [comment]
About update, do we have a way to update rom entity with dict ? 

We can easily to create a new entity with below 
m = SomeModel(**init_dict)

If we can update it with below, that will be awesome
m.update(**upd_dict)

Anyway, list and dict are core data format to communicate between python core and other modules. I believe it will be helpful. 

@josiahcarlson any thoughts?

**Comment by josiahcarlson** [comment]
Note that order-specific handling *may* only work consistently in Python 3.6+ , so I'm only testing that for Python 3.6 (right now, will do 3.7 in the future). These features are available in rom version 0.41.2 and later.

I went ahead and added `Model.update(*args, **kwargs)`, docs here: https://josiahcarlson.github.io/rom/rom.html#rom.Model.update

Docs for `Query.delete()` are here: https://josiahcarlson.github.io/rom/rom.html#rom.Query.delete

Docs for `session.delete()` are here: https://josiahcarlson.github.io/rom/rom.html#rom.util.Session.delete

I'm going to go ahead and close this, because it seems that I've got most/all of the easy fixes here. If you've got more requests, please feel free to create another issue. Thank you for your feedback and requests!

**Comment by 61t** [comment]
Awesome work @josiahcarlson  Thank you so much and have a wonderful day. :)
